includes/validation/class-amp-validation-manager.php

```diff
@@ -1617,19 +1594,56 @@ public static function get_amp_validate_nonce() {
 	}
 
 	/**
-	 * Whether to validate the front end response.
+	 * Whether the request is for validation data for a given URL.
+	 *
+	 * All AMP responses get validated, but when the amp_validate query parameter is present, then the source information
+	 * for each validation error is captured and the validation results are returned as JSON instead of the AMP HTML page.
 	 *
-	 * @return boolean Whether to validate.
+	 * @return bool|WP_Error Whether to validate. False is returned if it is not a validation request. WP_Error returned
+	 *                       if unauthenticated, unauthorized, and/or invalid nonce supplied. True returned if
+	 *                       validation response should be served.
 	 */
 	public static function should_validate_response() {
 		if ( ! isset( $_GET[ self::VALIDATE_QUERY_VAR ] ) ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
 			return false;
 		}
-		if ( self::has_cap() ) {
-			return true;
-		}
+
 		$validate_key = wp_unslash( $_GET[ self::VALIDATE_QUERY_VAR ] ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
-		return self::get_amp_validate_nonce() === $validate_key;
+		if ( self::get_amp_validate_nonce() !== $validate_key ) {
+			return new WP_Error(
+				'http_request_failed',
+				__( 'Nonce authentication failed.', 'amp' )
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Get validation response data.
+	 *
+	 * @return array Validation response data.
+	 */
+	public static function get_validation_response_data() {
+		$data = [
+			'results' => self::$validation_results,
+		];
+		if ( get_queried_object() ) {
```

Comment:
By pulling in global data you mean populating properties of the global `$wp_query`? Note that `\WP_Query::get_queried_object()` will short-circuit when `\WP_Query::$queried_object` has been set.

Reply:
Updated in 2a1dc540.

Reply:
Yes, but you're going through `get_queried_object()` first, and that fetches the global `$wp_query` object first. Although, unlikely, this is technically brittle code, as an operation in-between could change the object that variable points to, and you'll end up with a different queried object.

You're assuming that you'll get the same object every time, instead of ensuring you're only working with a single one yourself. (see https://ocramius.github.io/extremely-defensive-php/#/71)